### PR TITLE
Optimized string pools

### LINKED ISSUE #27: String Pooling
**Describe the bug**

ORC uses a lot of memory. One suggestion is to pool strings.

The interned string object:
* never moves in memory
* encodes the length, hash, and string data (probably just packed into uint8_t)
* is guaranteed to be unique
The reference to the interned string:
* Is only the size of a pointer (which points to the string, but can access the size & hash)
* Pointer equality is the same as string equality (they are unique)
* string_view can be easily and efficiently generated when needed (the length and data is encoded in the interned data).
* Reading a string is generally only 1-2 cache fetches
* Reading doesn't require synchronization (they are unique)
So "only" writing needs to be locked. That's still a big deal, mind you. But if we work out the threading performance, I think the interning is an optimal memory scheme without resorting to compression. (Ropes, tris, etc.) Strings are in coherent memory, packed, and unique. The instance of the string is always the size of a pointer.

### PR DESCRIPTION
**Pull request by leethomason**
Optimizes (for memory) the string pools. Also improves run-time performance as a side effect.

## Description

* `string_pool` is reduced to the size of one pointer
* strings are packed into contiguous memory
* `die` uses `string_pool` instead of std::string for the `_path`

The pool of strings is allocated per-thread. This improves both memory usage and performance:
```
            Mem(GB)     Time(sec)
main             83          63
string-pool      54          40   
```
Improving overall memory I find somewhat counter-intuitive. 
    * the highly used `struct die` is much smaller. (`pool_string` is now one pointer instead of 3, and a `std::string` became a `pool_string`)
    * the threads have good coherency; a given thread is generally processing a compilation unit, and the same strings repeat
    * and (in both solutions) the memory to store strings is greatly reduced, since they are stored in un-aligned blocks of packed memory

That performance increases is clearer: ORC uses a lot of memory, and it goes faster when it can use less of it.

## Related Issue

Fixes #27 

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by fosterbrereton** [comment]
Thanks for this work, Lee. I think I understand where you're trying to go with this. I am a little concerned about tucking everything under a pointer, and doing manual work to get e.g., the size and the hash out from under that pointer. (For example, it can make those values difficult to debug and/or maintain.) As an alternative, could `pool_string` be a data structure that is allowed to be bigger than a pointer, but then in the die we store a pointer to a pool_string?

In other words, could `die` change to be:

```c++
class die {
    // ...
    const pool_string* _object_file;
    const pool_string* _path;
    // ...
};
```
And then `empool` returns a `const pool_string*` instead?

**Comment by leethomason** [comment]
My thinking on that point (up for discussion, of course):
* `string` points to heap memory it manages
* `string_view` points to immutable (for the life of the `string_view`) memory of char*
* `pool_string` points to immutable memory with metadata

That's the first point, in that `pool_string` should "work like" `string_view`. They are used the same way, and in fact the only utility method on `pool_string` is to get a `string_view`.

The 2nd problem is technical; if `pool_string` is a struct, I expect it to obey alignment rules, which `pool_string` does not. The "structure" is packed and doesn't start on a 16-byte (or whatever) boundary. And we probably want to pack more in a future PR: we could use a 24-bit size and a 48-bit hash, for instance. I think the compilers we currently use could deal with the current packing and un-aligned reads, but it certainly can bite you.

To help with the brittleness of the solution:
* the getters for size() and hash() are in one place (no duplication)
* the empool is in one place (no duplication)
* if you look at the string_pool in a debugger you should see a (C) string

That's my thinking, but I'm balancing brittleness vs. compression, and there could be a better tradeoff.

**Comment by leethomason** [comment]
One idea: I could wrap up the empool(), and the getters() into a `string_pool` class so that all the weird code is in one place.

File: src/string_pool.cpp [hunk 1]
```diff
@@ -14,99 +14,87 @@
...
+    const char* empool(std::string_view incoming) {
         constexpr auto default_min_k = 16 * 1024 * 1024; // 16MB
-        const auto sz = incoming.size();
-        const auto tsz = sz + 1;
+        const uint32_t sz = (uint32_t)incoming.size();
+        const uint32_t tsz = sz + sizeof(uint32_t) + sizeof(size_t) + 1;
+        
         if (_n < tsz) {
             _n = std::max<std::size_t>(default_min_k, tsz);
+            _total_alloc += _n;
```
Comment (fosterbrereton) [inline src/string_pool.cpp:45]:
```suggestion
```

File: src/string_pool.cpp [hunk 2]
```diff
@@ -14,99 +14,87 @@
...
 
 /**************************************************************************************************/
 
+
 struct pool {
-    char* _p;
+    char* _p{nullptr};
     std::size_t _n{0};
     std::vector<std::unique_ptr<char[]>> _ponds;
+    std::size_t _total_alloc = 0;
```
Comment (fosterbrereton) [inline src/string_pool.cpp:36]: This looks unused; can it be removed?
```suggestion
```

File: include/orc/dwarf_structs.hpp [hunk 3]
```diff
@@ -68,7 +68,7 @@ struct attribute_value {
 
     void string(pool_string x) {
         _type |= type::string;
-        _string = std::move(x);
+        _string = x;
```
Comment (fosterbrereton) [inline include/orc/dwarf_structs.hpp:71]: Did the `move` here break things? Or is it just not necessary anymore?

File: include/orc/string_pool.hpp [hunk 4]
```diff
@@ -16,39 +16,107 @@
 
 struct pool_string;
 
+/*
+    Add a string to the pool. This is not thread safe.
```
Comment (fosterbrereton) [inline include/orc/string_pool.hpp:20]: Why is this operation not thread safe anymore?

**Comment by fosterbrereton** [comment]
> One idea: I could wrap up the empool(), and the getters() into a `string_pool` class so that all the weird code is in one place.

Thank you for the explanation, that does help a lot. I think I'd prefer to have all the weird code bundled into the source file, so then at least it's explicitly an implementation detail.

File: src/string_pool.cpp [hunk 5]
```diff
@@ -14,99 +14,87 @@
...
 
 pool_string empool(std::string_view src) {
-    auto& s = set();
-    pool_string needle(src);
-    auto found = s.find(needle);
-
-    if (found != s.end()) {
-        return *found;
+    if (src.empty())
+        return pool_string(nullptr);
```
[the remark this answers is not in the document]
Comment (leethomason) [reply src/string_pool.cpp:90]: good catch

File: src/string_pool.cpp (hunk 1 above)
Comment (leethomason) [reply src/string_pool.cpp:49]: thanks!

File: src/string_pool.cpp (hunk 2 above)
Comment (leethomason) [reply src/string_pool.cpp:40]: it can, thanks!

File: include/orc/string_pool.hpp (hunk 4 above)
Comment (leethomason) [reply include/orc/string_pool.hpp]: good catch, again! old comment.

**Comment by leethomason** [comment]
General cleanup. Tried to keep the weird packing parts in the same file close together, at least. Improved comments.

File: src/string_pool.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (leethomason) [reply src/string_pool.cpp:88]: turns out to not quite be correct - a string_view can be empty (and a default is returned) but a pool_string is defined so that _data=nullptr is empty, and anything else must have content.

File: include/orc/dwarf_structs.hpp (hunk 3 above)
Comment (leethomason) [reply include/orc/dwarf_structs.hpp:71]: Just not necessary; the pool_string is just a pointer. (From a style / communication perspective, I'm not sure which is better.)

Comment (fosterbrereton) [reply include/orc/dwarf_structs.hpp:71]: It's just a pointer _for now_ 😃

What about using pass-by-rvalue-ref here, and then peppering the code with `copy` / `move` as appropriate? It'd probably take this PR in a direction you don't want it to go. In which case, can we track it as an issue?

File: src/string_pool.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (fosterbrereton) [reply src/string_pool.cpp:88]: Can you add that detail to the comments in the string pool header, then, or here? I could see someone trying to optimize this line later and breaking intent.

File: include/orc/string_pool.hpp [hunk 6]
```diff
@@ -16,40 +16,81 @@
...
-    std::string allocate_string() const { return std::string(_s); }
-    std::filesystem::path allocate_path() const { return std::filesystem::path(_s); }
+    std::string_view view() const {
+        if (!_data) return default_view;
+        std::uint32_t size = get_size(_data);
+        return std::string_view(_data, size);
+    }
+    
+    std::string allocate_string() const { 
+        assert(_data);
```
Comment (fosterbrereton) [inline include/orc/string_pool.hpp:56]: This assert isn't necessary anymore - a `pool_string` with no data can still create a valid `std::string`.

File: include/orc/string_pool.hpp [hunk 7]
```diff
@@ -16,40 +16,81 @@
...
+        return std::string(view()); 
+    }
+    
+    std::filesystem::path allocate_path() const { 
+        assert(_data);
```
Comment (fosterbrereton) [inline include/orc/string_pool.hpp:61]: This assert isn't necessary anymore - a `pool_string` with no data can still create a valid `std::filesystem::path`.

File: include/orc/string_pool.hpp [hunk 8]
```diff
@@ -16,40 +16,81 @@
...
+    size_t hash() const {
+        if (!_data) return 0;
+        return get_hash(_data);
+    }
 
     friend inline bool operator==(const pool_string& x, const pool_string& y) {
-        return x._h == y._h;
+        if (x._data == y._data)
+            return true;
+        return x.view() == y.view();
```
Comment (fosterbrereton) [inline include/orc/string_pool.hpp:73]: Easier to read?
```suggestion
        return x._data == y._data || x.view() == y.view();
```

File: src/string_pool.cpp [hunk 9]
```diff
@@ -14,99 +14,101 @@
...
 
-auto& pool() {
-#if ORC_FEATURE(LEAKY_MEMORY)
-    thread_local struct pool& instance = *(new struct pool());
-#else
-    thread_local struct pool instance;
-#endif // ORC_FEATURE(LEAKY_MEMORY)
-
-    return instance;
+std::size_t pool_string::get_size(const char* d) {
```
Comment (fosterbrereton) [inline src/string_pool.cpp:68]: Thanks for hiding these 😃

**Review by fosterbrereton — APPROVED** [review]
LGTM! A couple minor nits. Thanks for getting the memory footprint down, Lee!

File: include/orc/dwarf_structs.hpp (hunk 3 above)
Comment (leethomason) [reply include/orc/dwarf_structs.hpp:71]: I assert it needs to be a pointer forever, else it isn't an interned string, and breaks the optimization. I did add:
```
static_assert(sizeof(pool_string) <= sizeof(intptr_t), "pool_string is design to be as small and fast to copy as a pointer.");
```
There's an argument that `pool_string` should be a typedef of a pointer (which makes the constraint clearer) but it is very useful to decorate it with methods.

File: include/orc/string_pool.hpp (hunk 6 above)
Comment (leethomason) [reply include/orc/string_pool.hpp:56]: good catch - left over from earlier version.

File: src/string_pool.cpp (hunk 9 above)
Comment (leethomason) [reply src/string_pool.cpp:68]: np - they are weird

File: src/string_pool.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (leethomason) [reply src/string_pool.cpp:91]: Added comment
